design/proposals/module-linking/Binary.md

```diff
@@ -1,198 +1,174 @@
+* For outer aliases, `ct` is validated to be *less than* the number of enclosing
+  modules and `li` is validated to be a valid index in the `def-kind` index space.
+* For the time being, export aliases may not have the `type` `def-kind` and outer
+  aliases may *only* have the `module` or `type` `def-kind`s.
 
-**Validation**
-
-* The instance, module and type indices are all in bounds. Remember that "in bounds"
-  means only those definitions preceding this alias definition in binary format
-  order. In the case of outer aliases, this means the position of the nested module
-  definition in the outer module.
-* Aliased instance export names must be found in `i`'s instance type with a
-  matching definition kind.
-* The `ct` of an outer alias must be *less than* the number of enclosing modules
-  which implicitly prevents outer aliases from appearing in top-level modules.
-
-## Function section
-
-**Validation**
-
-* Type indices must point to function types.
 
-## Export section
-
-update
-[`exportdesc`](https://webassembly.github.io/spec/core/binary/modules.html#binary-exportdesc)
+## Type Definitions
 
+(See [type definitions](Explainer.md#type-definitions) in the explainer.)
 ```
-exportdesc ::=
-    ...
-    0x05 x:moduleidx                        -> module x
-    0x06 x:instanceidx                      -> instance x
+type             ::= 0x7f e*:vec(export-decl)               -> (instance e*)
+                   | 0x7e mtd*:vec(module-type-decl)        -> (module mtd*)
+                   | 0x7d p*:vec(val-type) r*:vec(val-type) -> (func (param p)* (result r)*)
+export-decl      ::= nm:<name> dt:<def-type>                -> (export nm dt)
+module-type-decl ::= 0x00 t:<type>                          -> t
+                   | 0x01 i:<import>                        -> i
+                   | 0x05 a:<alias>                         -> a
+                   | 0x06 e:<export-decl>                   -> e
+val-type         ::= 0x00 vt:<core:valtype>                 -> vt
 ```
-
-**Validation**
-
-* Module/instance indexes must be in-bounds.
-
+Notes:
+* Modules types create a fresh type index space that is populated and
+  referenced by their `module-type-decl`s. Thus, for a module type that imports
+  a function, the import `module-type-decl` must be preceded by either a type
+  or alias `module-type-decl` that adds the function type to the type index
+  space.
+* Currently, the only allowed form of `alias` in a `module-type-decl` is
+  an `(alias outer ct li (type))`. In the future, other kinds of aliases
+  will be needed and this restriction will be relaxed.
+* To avoid redefining the whole binary format of Core WebAssembly's [`valtype`]
+  and to allow implementation reuse, the binary format for adapter modules'
+  `val-type` embeds `core:valtype`. A `0x00` prefix is used so that future
+  additions to `core:valtype` cannot conflict with future additions to
+  `val-type` (viz., in Interface Types).
```

Comment:
Hm, I'd much prefer to make the module type opcodes a proper extension of the core ones.

Reply:
As discussed offline, will keep the current split.